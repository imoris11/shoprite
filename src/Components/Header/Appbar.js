import React from 'react';
import PropTypes from 'prop-types';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import InputBase from '@material-ui/core/InputBase';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import { fade } from '@material-ui/core/styles/colorManipulator';
import { withStyles } from '@material-ui/core/styles';
import MenuIcon from '@material-ui/icons/Menu';
import SearchIcon from '@material-ui/icons/Search';
import AccountCircle from '@material-ui/icons/AccountCircle';
import '../../Resources/css/header.css';
import { Link } from 'react-router-dom';

const styles = theme => ({
  root: {
    width: '100%',
  },
  grow: {
    flexGrow: 1,
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20,
  },
  title: {
    display: 'none',
    [theme.breakpoints.up('sm')]: {
      display: 'block',
    },
  },
  search: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginRight: theme.spacing.unit * 2,
    marginLeft: 0,
    width: '100%',
    [theme.breakpoints.up('sm')]: {
      marginLeft: theme.spacing.unit * 3,
      width: 'auto',
    },
  },
  searchIcon: {
    width: theme.spacing.unit * 9,
    height: '100%',
    position: 'absolute',
    pointerEvents: 'none',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  inputRoot: {
    color: 'inherit',
    width: '100%',
  },
  inputInput: {
    paddingTop: theme.spacing.unit,
    paddingRight: theme.spacing.unit,
    paddingBottom: theme.spacing.unit,
    paddingLeft: theme.spacing.unit * 10,
    transition: theme.transitions.create('width'),
    width: '100%',
    [theme.breakpoints.up('md')]: {
      width: 200,
    },
  },
  sectionDesktop: {
    display: 'none',
    [theme.breakpoints.up('md')]: {
      display: 'flex',
    },
  },
  sectionMobile: {
    display: 'flex',
    [theme.breakpoints.up('md')]: {
      display: 'none',
    },
  },
  menuItems:{
    margin:10,
    padding:5
  },
});

class PrimarySearchAppBar extends React.Component {
  constructor () {
    super()
    this.state = {
      anchorEl: null,
      mobileMoreAnchorEl: null,
      departments:[],
      token:'',
      searchString:''
    }
    this.departments = []
  }
  componentDidMount () {
    this.fetchData()
    this.confirmUserStatus()
  }

  fetchData = () => {
    fetch("https://backendapi.turing.com/departments").then(response => response.json()).then(results => {
      if (!results.error) {
        this.setState({departments:results})
      }
    })
  }

  async confirmUserStatus () {
    let token = await localStorage.getItem('token');
    if (token) {
      this.setState({token, loggedIn:true})
    }
  }

  handleProfileMenuOpen = event => {
    this.setState({ anchorEl: event.currentTarget });
  };

  handleMenuClose = () => {
    this.setState({ anchorEl: null });
    this.handleMobileMenuClose();
  };

  handleMobileMenuOpen = event => {
    this.setState({ mobileMoreAnchorEl: event.currentTarget });
  };

  handleMobileMenuClose = () => {
    this.setState({ mobileMoreAnchorEl: null });
  };

  logout = () => {
    this.handleMobileMenuClose();
    localStorage.clear();
    window.location.href="/";
  }

  searchProducts = (e) =>  {
    e.preventDefault();
    //Temporary fix. Redirect from react-router-dom crashes
    window.location.href="/search/"+this.state.searchString;
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name] :e.target.value
    })
  }
  render() {
    let {departments} = this.state;
    const { anchorEl, mobileMoreAnchorEl } = this.state;
    const { classes } = this.props;
    const isMenuOpen = Boolean(anchorEl);
    const isMobileMenuOpen = Boolean(mobileMoreAnchorEl);

    const renderMenu = (
      <Menu
        anchorEl={anchorEl}
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        open={isMenuOpen}
        onClose={this.handleMenuClose}
      >
        <Link to='/profile'>
          <MenuItem onClick={this.handleMenuClose}>My account</MenuItem>
        </Link>
        <MenuItem onClick={this.logout}>Log Out</MenuItem>
      </Menu>
    );

    const renderMobileMenu = (
      <Menu
        anchorEl={mobileMoreAnchorEl}
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        open={isMobileMenuOpen}
        onClose={this.handleMenuClose}
      >
        {departments.map((department, key)=>
          <Link key={key} to={"/catalogue/" + department.department_id + "/" +department.name}>
            <MenuItem onClick={this.handleMobileMenuClose}>
              <Typography variant="h6" color="inherit" className={classes.menuItems}>
                {department.name}
              </Typography>
            </MenuItem>
          </Link>
        )}
      </Menu>
    );

    return (
      <div className={classes.root}>
        <AppBar position="static" style={{backgroundColor:'#f62f5e'}}>
          <Toolbar>
            <Link className='navbar-brand' to='/'>
              <Typography className={classes.title} variant="h6" color="inherit" noWrap>
                SHOPMATE
              </Typography>
            </Link>

            <div className={classes.grow} />
            <div className={classes.sectionDesktop}>
              {departments.map((department, key)=>
                <Link key={key} to={"/catalogue/" + department.department_id + "/" +department.name}>
                  <Typography variant="h6" color="inherit" className={classes.menuItems + " secondary-link"}>
                    {department.name}
                  </Typography>
                </Link>
              )}
            </div>
            <div className={classes.grow} />
            <form onSubmit={this.searchProducts}>
              <div className={classes.search}>
                <div className={classes.searchIcon}>
                  <SearchIcon />
                </div>
                <InputBase
                  placeholder="Searchâ€¦"
                  onChange={this.handleChange}
                  id={'searchBox'}
                  name='searchString'
                  value={this.state.searchString}
                  classes={{
                    root: classes.inputRoot,
                    input: classes.inputInput,
                  }}
                />
              </div>
              <button id="submitBtn" type="submit" style={{display:'none'}}></button>
            </form>
            <div>
              {this.state.loggedIn && (
            <div>
              <IconButton
                aria-owns={isMenuOpen ? 'menu-appbar' : undefined}
                aria-haspopup="true"
                onClick={this.handleProfileMenuOpen}
                color="inherit"
              >
                <AccountCircle />
              </IconButton>
            </div>
          )}
            </div>
            <div className={classes.sectionMobile}>
              <IconButton aria-haspopup="true" onClick={this.handleMobileMenuOpen} color="inherit">
                <MenuIcon />
              </IconButton>
            </div>
          </Toolbar>
        </AppBar>
        {renderMenu}
        {renderMobileMenu}
      </div>
    );
  }
}

PrimarySearchAppBar.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(PrimarySearchAppBar);
