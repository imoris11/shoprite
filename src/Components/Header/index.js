import React, {Component} from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import CartIcon from '@material-ui/icons/ShoppingCart';
import ArrowDropUp from '@material-ui/icons/ArrowDropUp';
import Badge from '@material-ui/core/Badge';
import AppBar from './Appbar';
import {Row, Col} from 'react-bootstrap';
import {Link} from 'react-router-dom';
import Registration from '../Registration';
import CartModal from '../Cart/Cart-modal';
import '../../Resources/css/header.css';
const styles = theme => ({
  root: {
    flexGrow: 1,
  },
  grow: {
    flexGrow: 1,
  },
  menuButton: {
    marginLeft: -12,
    marginRight: 20,
  },
  menuItems:{
    marginTop:10,
  },
  margin: {
    margin: theme.spacing.unit * 2,
  },
});

class Header extends Component{
  constructor (props) {
    super(props);
    this.state = {
      showDialog:false,
      badgeValue:0,
      showCart:false
    }
    this.handleScroll = this.handleScroll.bind(this);
    this.localStorageUpdated = this.localStorageUpdated.bind(this);
  }
  componentDidMount() {
    //Add listener to show scroll to top UI
    window.addEventListener("scroll", this.handleScroll);
    //Add listener to localStorage on document
    this.addLocalStorageListener();
    //Get last stored number of products in cart
    this.getLocalBadgeValue();
    //Confirm if user is logged in
    this.confirmUserStatus();
  }
  addLocalStorageListener () {
    var originalSetItem = localStorage.setItem;
    localStorage.setItem = function() {
      var event = new Event('itemInserted');
      document.dispatchEvent(event);
      originalSetItem.apply(this, arguments);
    }
    document.addEventListener("itemInserted", this.localStorageUpdated, false);
  }
  confirmUserStatus () {
    let token = localStorage.getItem('token');
    let user = localStorage.getItem('customer');
    if (token) {
      user = JSON.parse(user);
      this.setState({token, loggedIn:true, name: user.firstName + " " + user.lastName});
    }
  }
  getLocalBadgeValue () {
    let badgeValue = localStorage.getItem('badge');
    if (badgeValue) {
      this.setState({badgeValue});
    }
  }
  componentWillUnmount() {
    window.removeEventListener("scroll", this.handleScroll);
    window.removeEventListener('itemInserted', this.localStorageUpdated);
  }
  localStorageUpdated(e){
    let badgeValue = localStorage.getItem('badge');
    if (badgeValue && Number(badgeValue) !== this.state.badgeValue) {
      this.setState({badgeValue});
    }else if (!badgeValue){
      this.setState({badgeValue});
    }
  }
  toggleDialog = () => {
    this.setState(prevState => ({showDialog:true}))
  }
  toggleCartModal = () => {
    this.setState(prevState => ({showCart:true}))
  }
  handleClose = () => {
    this.setState(prevState => ({showCart:false, showDialog:false}))
  }
  handleScroll() {
    const windowHeight = "innerHeight" in window ? window.innerHeight : document.documentElement.offsetHeight;
    const body = document.body;
    const html = document.documentElement;
    const docHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight,  html.scrollHeight, html.offsetHeight);
    const windowBottom = windowHeight + window.pageYOffset;
    if (windowBottom >= docHeight) {
      this.setState({
        isAtBottom:true
      });
    }
  }
  scrollToTop = () => {
    window.scrollTo(0, 0);
  }
  render () {
    const { classes } = this.props;
    return (
      <div className={classes.root}>
        <Col style={{backgroundColor:'#2e2e2e', color:'white'}}>
          <Row>
            <Col sm={3} xs={10}>
              {!this.state.token ?
                 <p style={{marginTop:10}}>Hi! <span className='span-link' onClick={this.toggleDialog}>Sign in</span>
                 &nbsp; Or <span className='span-link' onClick={this.toggleDialog}>Register</span>
               </p> : <p style={{marginTop:10}}>Hi! <span className='span-link'>{this.state.name}</span></p>
             }
            </Col>
            <Col sm={8} className='d-none d-sm-block'>
              <Row>
                <Col md={2} sm={3}>
                  <Link to="/deals" className='primary-link'>
                  <p className={classes.menuItems}>Daily Deals</p></Link>
                </Col>
                <Col md={1} sm={1}>
                  <Link to="/sell" className='primary-link' >
                  <p className={classes.menuItems} style={{marginTop:10}}>Sell</p>
                  </Link>
                </Col>
                <Col md={3} sm={3}>
                  <Link to="/contact" className='primary-link' >
                  <p className={classes.menuItems} style={{marginTop:10}}>Help & Contact</p>
                </Link>
                </Col>
                <Col md={6} sm={5}></Col>
              </Row>
            </Col>
            <Col sm={1} xs={2}>
              <IconButton onClick={this.toggleCartModal} className={classes.menuButton} color="inherit" aria-label="Menu">
                <Badge badgeContent={this.state.badgeValue} color="primary">
                  <CartIcon />
                </Badge>
            </IconButton>
          </Col>
          </Row>
        </Col>
        <AppBar />
        {this.state.isAtBottom && <div onClick={this.scrollToTop} className='text-center floating-button'>
          <ArrowDropUp className='floating-icon' style={{fontSize:'50px'}} />
        </div> }
        <Registration showDialog={this.state.showDialog} handleClose={this.handleClose} />
        <CartModal showDialog={this.state.showCart} handleClose={this.handleClose} items={this.state.badgeValue} />
      </div>
    );
  }
}

Header.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Header);
