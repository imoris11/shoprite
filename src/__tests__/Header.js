import React from 'react';
import ReactDOM from 'react-dom';
import { mount, shallow } from 'enzyme';
import Header from '../Components/Header';

describe('Test Header Component', () => {

   it('renders without crashing', () => {
      shallow(<Header />);
    });

    it('renders correctly', ()=> {
      const header = shallow(<Header />);
      expect(header).toMatchSnapshot();
    });

});
