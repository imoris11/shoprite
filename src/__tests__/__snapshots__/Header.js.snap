// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Header Component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(Header) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "classes": Object {
        "grow": "Header-grow-2",
        "margin": "Header-margin-5",
        "menuButton": "Header-menuButton-3",
        "menuItems": "Header-menuItems-4",
        "root": "Header-root-1",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "classes": Object {
          "grow": "Header-grow-2",
          "margin": "Header-margin-5",
          "menuButton": "Header-menuButton-3",
          "menuItems": "Header-menuItems-4",
          "root": "Header-root-1",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
